File: android/app/src/main/java/com/openburnbar/data/stores/MemoryWalletStore.kt
package com.openburnbar.data.stores

import androidx.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.openburnbar.data.firebase.FirestoreRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Live Memory Boost wallet cache. Cloud Functions are the sole writer;
 * this listener only reads `users/{uid}/memoryWallet/current`.
 */
class MemoryWalletStore(
    initialFirestore: FirebaseFirestore? = null,
    initialFirebaseAuth: FirebaseAuth? = null,
) : ViewModel() {
    private val firestore = initialFirestore ?: FirestoreRepository.database()
    private val firebaseAuth = initialFirebaseAuth ?: FirebaseAuth.getInstance()

    private val _textTokens = MutableStateFlow(0L)
    val textTokens: StateFlow<Long> = _textTokens.asStateFlow()

    private val _multimodalTokens = MutableStateFlow(0L)
    val multimodalTokens: StateFlow<Long> = _multimodalTokens.asStateFlow()

    private val _pendingTextTokens = MutableStateFlow(0L)
    val pendingTextTokens: StateFlow<Long> = _pendingTextTokens.asStateFlow()

    private val _pendingMultimodalTokens = MutableStateFlow(0L)
    val pendingMultimodalTokens: StateFlow<Long> = _pendingMultimodalTokens.asStateFlow()

    private val _loadFailed = MutableStateFlow(false)
    val loadFailed: StateFlow<Boolean> = _loadFailed.asStateFlow()

    private var authListener: FirebaseAuth.AuthStateListener? = null
    private var walletListener: ListenerRegistration? = null

    fun start() {
        if (authListener != null) return
        val listener =
            FirebaseAuth.AuthStateListener { auth ->
                listen(auth.currentUser?.uid)
            }
        firebaseAuth.addAuthStateListener(listener)
        authListener = listener
        listen(firebaseAuth.currentUser?.uid)
    }

    fun stop() {
        walletListener?.remove()
        walletListener = null
        authListener?.let { firebaseAuth.removeAuthStateListener(it) }
        authListener = null
        resetBalances()
        _loadFailed.value = false
    }

    override fun onCleared() {
        stop()
        super.onCleared()
    }

    private fun listen(uid: String?) {
        walletListener?.remove()
        walletListener = null
        if (uid == null) {
            resetBalances()
            _loadFailed.value = false
            return
        }
        walletListener =
            firestore.collection("users")
                .document(uid)
                .collection("memoryWallet")
                .document("current")
                .addSnapshotListener { snap, error ->
                    if (error != null) {
                        _loadFailed.value = true
                        return@addSnapshotListener
                    }
                    _loadFailed.value = false
                    val data = snap?.data
                    _textTokens.value = longValue(data?.get("textTokens"))
                    _multimodalTokens.value = longValue(data?.get("multimodalTokens"))
                    _pendingTextTokens.value = longValue(data?.get("pendingTextTokens"))
                    _pendingMultimodalTokens.value = longValue(data?.get("pendingMultimodalTokens"))
                }
    }

    private fun resetBalances() {
        _textTokens.value = 0
        _multimodalTokens.value = 0
        _pendingTextTokens.value = 0
        _pendingMultimodalTokens.value = 0
    }

    private fun longValue(raw: Any?): Long {
        return when (raw) {
            is Number -> raw.toLong()
            else -> 0L
        }
    }
}

File: OpenBurnBarMobile/Services/MemoryWalletStore.swift
import Foundation
@preconcurrency import FirebaseAuth
import FirebaseCore
@preconcurrency import FirebaseFirestore

/// Live Memory Boost wallet cache. Cloud Functions are the sole writer;
/// this listener only reads `users/{uid}/memoryWallet/current`.
@MainActor
@Observable
final class MemoryWalletStore {
    private(set) var textTokens: Int = 0
    private(set) var multimodalTokens: Int = 0
    private(set) var pendingTextTokens: Int = 0
    private(set) var pendingMultimodalTokens: Int = 0
    private(set) var loadFailed: Bool = false

    @ObservationIgnored private var listener: ListenerRegistration?
    @ObservationIgnored private nonisolated(unsafe) var authHandle: AuthStateDidChangeListenerHandle?

    func start() {
        guard authHandle == nil else { return }
        authHandle = Auth.auth().addStateDidChangeListener { [weak self] _, user in
            Task { @MainActor in
                self?.listen(uid: user?.uid)
            }
        }
        listen(uid: Auth.auth().currentUser?.uid)
    }

    func stop() {
        listener?.remove()
        listener = nil
        if let authHandle {
            Auth.auth().removeStateDidChangeListener(authHandle)
            self.authHandle = nil
        }
        resetBalances()
        loadFailed = false
    }

    private func listen(uid: String?) {
        listener?.remove()
        listener = nil
        guard let uid, FirebaseApp.app() != nil else {
            resetBalances()
            loadFailed = false
            return
        }
        listener = FirestoreRepository.database
            .document("users/\(uid)/memoryWallet/current")
            .addSnapshotListener { [weak self] snap, error in
                Task { @MainActor in
                    guard let self else { return }
                    if error != nil {
                        self.loadFailed = true
                        return
                    }
                    self.loadFailed = false
                    let data = snap?.data()
                    self.textTokens = Self.intValue(data?["textTokens"])
                    self.multimodalTokens = Self.intValue(data?["multimodalTokens"])
                    self.pendingTextTokens = Self.intValue(data?["pendingTextTokens"])
                    self.pendingMultimodalTokens = Self.intValue(data?["pendingMultimodalTokens"])
                }
            }
    }

    private func resetBalances() {
        textTokens = 0
        multimodalTokens = 0
        pendingTextTokens = 0
        pendingMultimodalTokens = 0
    }

    private static func intValue(_ raw: Any?) -> Int {
        if let value = raw as? Int { return value }
        if let value = raw as? Int64 { return Int(value) }
        if let value = raw as? Double { return Int(value) }
        return 0
    }
}
